# 4.9 Task PART1

## This Notebook contains: 
1. Import libraries and data
2. Data wrangling
3. Data consistency check
4. Data merge
5. Export data

# 01. Import Libraries and Data

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Set folder path
path = r'C:\Users\hachl\Downloads'

In [3]:
# Import the customer data set
customer_df = pd.read_csv(os.path.join(path,'Original Data','customers.csv'))

In [4]:
# Check dataframe
customer_df.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
customer_df.shape

(206209, 10)

# 02. Data Wrangling

### Columns that need renaming are 'First Name','Surnam','Gender','STATE','Age','n_dependants',fam_status'. I will not drop any columns for now as I will probably need all in Part2.

In [6]:
# Rename variables

customer_df.rename(columns = {'First Name' : 'first_name'}, inplace = True)
customer_df.rename(columns = {'Surnam' : 'last_name'}, inplace = True)
customer_df.rename(columns = {'Gender' : 'gender'}, inplace = True)
customer_df.rename(columns = {'STATE' : 'state'}, inplace = True)
customer_df.rename(columns = {'Age' : 'age'}, inplace = True)
customer_df.rename(columns = {'n_dependants' : 'number_of_dependents'}, inplace = True)
customer_df.rename(columns = {'fam_status' : 'marital_status'}, inplace = True)

In [7]:
# Check output
customer_df.columns

Index(['user_id', 'first_name', 'last_name', 'gender', 'state', 'age',
       'date_joined', 'number_of_dependents', 'marital_status', 'income'],
      dtype='object')

# 03. Data consistency checks

In [8]:
# Search for missing values in all variables

customer_df['user_id'].value_counts(dropna = False)

user_id
26711     1
67322     1
173044    1
61044     1
98344     1
         ..
146847    1
154991    1
172193    1
184326    1
80148     1
Name: count, Length: 206209, dtype: int64

In [9]:
customer_df['first_name'].value_counts(dropna = False)

first_name
NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: count, Length: 208, dtype: int64

In [10]:
customer_df['last_name'].value_counts(dropna = False)

last_name
Hamilton      252
Randall       248
Lamb          243
Pennington    243
Barnett       242
             ... 
Poole         172
Bauer         166
Pearson       164
Payne         163
Jordan        162
Name: count, Length: 1000, dtype: int64

In [11]:
customer_df['gender'].value_counts(dropna = False)

gender
Male      104067
Female    102142
Name: count, dtype: int64

In [12]:
customer_df['state'].value_counts(dropna = False)

state
Florida                 4044
Colorado                4044
Illinois                4044
Alabama                 4044
District of Columbia    4044
Hawaii                  4044
Arizona                 4044
Connecticut             4044
California              4044
Indiana                 4044
Arkansas                4044
Alaska                  4044
Delaware                4044
Iowa                    4044
Idaho                   4044
Georgia                 4044
Wyoming                 4043
Mississippi             4043
Oklahoma                4043
Utah                    4043
New Hampshire           4043
Kentucky                4043
Maryland                4043
Rhode Island            4043
Massachusetts           4043
Michigan                4043
New Jersey              4043
Kansas                  4043
South Dakota            4043
Minnesota               4043
Tennessee               4043
New York                4043
Washington              4043
Louisiana               4043
Montana 

In [13]:
customer_df['age'].value_counts(dropna = False)

age
19    3329
55    3317
51    3317
56    3306
32    3305
      ... 
65    3145
25    3127
66    3114
50    3102
36    3101
Name: count, Length: 64, dtype: int64

In [14]:
customer_df['date_joined'].value_counts(dropna = False)

date_joined
9/17/2018     213
2/10/2018     212
4/1/2019      211
9/21/2019     211
12/19/2017    210
             ... 
9/1/2018      141
1/22/2018     140
11/24/2017    139
7/18/2019     138
8/6/2018      128
Name: count, Length: 1187, dtype: int64

In [15]:
customer_df['number_of_dependents'].value_counts(dropna = False)

number_of_dependents
0    51602
3    51594
1    51531
2    51482
Name: count, dtype: int64

In [16]:
customer_df['marital_status'].value_counts(dropna = False)

marital_status
married                             144906
single                               33962
divorced/widowed                     17640
living with parents and siblings      9701
Name: count, dtype: int64

In [17]:
customer_df['income'].value_counts(dropna = False)

income
57192     10
95891     10
95710     10
97532      9
98675      9
          ..
73141      1
71524      1
74408      1
44780      1
148828     1
Name: count, Length: 108012, dtype: int64

### The only missing data is within 'first_name' column (NaN: 11259) but both user_id and last name are available to trace the customer, so we cannot drop the rows. I will fill in these missing values with 'unknown' for clarity.

In [18]:
# Replace missing values

customer_df['first_name'].fillna('unknown', inplace=True)

In [19]:
# Search for duplicates: create subset containing only duplicate rows

df_dups = customer_df[customer_df.duplicated()]

In [20]:
# Check output
df_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, number_of_dependents, marital_status, income]
Index: []

### Dataframe is empty so there are no duplicate rows.

In [21]:
# Check frequency table of customer df for anomalies --> there seem to be none

customer_df.describe()

user_id            age  number_of_dependents         income
count  206209.000000  206209.000000         206209.000000  206209.000000
mean   103105.000000      49.501646              1.499823   94632.852548
std     59527.555167      18.480962              1.118433   42473.786988
min         1.000000      18.000000              0.000000   25903.000000
25%     51553.000000      33.000000              0.000000   59874.000000
50%    103105.000000      49.000000              1.000000   93547.000000
75%    154657.000000      66.000000              3.000000  124244.000000
max    206209.000000      81.000000              3.000000  593901.000000

In [22]:
# Check for mixed-type data in customer df --> NO OUTPUT 

for col in customer_df.columns.tolist():
  weird = (customer_df[[col]].applymap(type) != customer_df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (customer_df[weird]) > 0:
    print (col)

In [23]:
# Check memory usage of customer data set
customer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  int64 
 1   first_name            206209 non-null  object
 2   last_name             206209 non-null  object
 3   gender                206209 non-null  object
 4   state                 206209 non-null  object
 5   age                   206209 non-null  int64 
 6   date_joined           206209 non-null  object
 7   number_of_dependents  206209 non-null  int64 
 8   marital_status        206209 non-null  object
 9   income                206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [24]:
# Change 'int64' data types to reduce memory usage

customer_df['user_id'] = customer_df['user_id'].astype('int32')
customer_df['age'] = customer_df['age'].astype('int8')
customer_df['number_of_dependents'] = customer_df['number_of_dependents'].astype('int8')
customer_df['income'] = customer_df['income'].astype('int32')

In [25]:
# Check if changes affected the data

customer_df.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependents marital_status  income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374

In [26]:
# Check memory usage again

customer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  int32 
 1   first_name            206209 non-null  object
 2   last_name             206209 non-null  object
 3   gender                206209 non-null  object
 4   state                 206209 non-null  object
 5   age                   206209 non-null  int8  
 6   date_joined           206209 non-null  object
 7   number_of_dependents  206209 non-null  int8  
 8   marital_status        206209 non-null  object
 9   income                206209 non-null  int32 
dtypes: int32(2), int8(2), object(6)
memory usage: 11.4+ MB


# 04. Data merge

In [3]:
# Import second data set

ords_prods_merge = pd.read_pickle(os.path.join(path, '4.3_orders_products','02 Data','Prepared Data', 'dataset_aggregated.pkl'))

In [4]:
# Check the second dataframe

ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  ...    busiest_day    busiest_days    busiest_hour  \
0         Soda  ...  Regularly day  Regularly busy  Average orders   
1         Soda  ...  Regularly day    Slowest days  Average orders   
2         Soda  ...  Regularly day    Slowest days     Most orders   
3         Soda  ...      Least day    Slowest days  Average orders   
4         Soda  ...      Least day    Slowest days     Most orders   

    busiest_hours max_order  loyalty_flag avg_price spending_flag median_dslo  \
0  Average orders        10  New customer  6.367797   Low spender        20.5   
1  Average orders        10  New customer  6.367797   Low spender        20.5   
2     Most orders        10  New customer  6.367797   Low spender        20.5   
3  Average orders        10  New customer  6.367797   Low spender        20.5   
4     Most orders        10  New customer  6.367797   Low spender        20.5   

          frequency_flag  
0  Non-frequent customer  
1  Non-frequent customer  
2  Non-frequent customer  
3  Non-frequent customer  
4  Non-frequent customer  

[5 rows x 25 columns]

In [5]:
# Replace missing values

ords_prods_merge['days_since_last_order'].fillna('0', inplace=True)

In [6]:
# Check shape of second df prior merging

ords_prods_merge.shape

(32404859, 25)

In [7]:
# Check memory usage too

ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 25 columns):
 #   Column                 Dtype   
---  ------                 -----   
 0   order_id               int32   
 1   user_id                int32   
 2   order_number           int32   
 3   orders_day_of_week     int8    
 4   order_hour_of_day      int8    
 5   days_since_last_order  object  
 6   product_id             int32   
 7   add_to_cart_order      int32   
 8   reordered              int32   
 9   product_name           object  
 10  aisle_id               int32   
 11  department_id          int32   
 12  prices                 float64 
 13  _merge                 category
 14  price_range_loc        object  
 15  busiest_day            object  
 16  busiest_days           object  
 17  busiest_hour           object  
 18  busiest_hours          object  
 19  max_order              int32   
 20  loyalty_flag           object  
 21  avg_price              float6

In [8]:
# Change data types to reduce memory usage

ords_prods_merge['order_number'] = ords_prods_merge['order_number'].astype('int8')
ords_prods_merge['days_since_last_order'] = ords_prods_merge['days_since_last_order'].astype('int8')
ords_prods_merge['add_to_cart_order'] = ords_prods_merge['add_to_cart_order'].astype('int8')
ords_prods_merge['reordered'] = ords_prods_merge['reordered'].astype('int8')
ords_prods_merge['prices'] = ords_prods_merge['reordered'].astype('float32')
ords_prods_merge['max_order'] = ords_prods_merge['max_order'].astype('int8')
ords_prods_merge['avg_price'] = ords_prods_merge['avg_price'].astype('float32')
ords_prods_merge['median_dslo'] = ords_prods_merge['median_dslo'].astype('float16')

In [9]:
# Check output

ords_prods_merge.head() # data doesn´t seem to have been corrupted

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                      0         196                  1          0   
1                     15         196                  1          1   
2                     21         196                  1          1   
3                     29         196                  1          1   
4                     28         196                  1          1   

  product_name  ...    busiest_day    busiest_days    busiest_hour  \
0         Soda  ...  Regularly day  Regularly busy  Average orders   
1         Soda  ...  Regularly day    Slowest days  Average orders   
2         Soda  ...  Regularly day    Slowest days     Most orders   
3         Soda  ...      Least day    Slowest days  Average orders   
4         Soda  ...      Least day    Slowest days     Most orders   

    busiest_hours max_order  loyalty_flag avg_price spending_flag median_dslo  \
0  Average orders        10  New customer  6.367796   Low spender        20.5   
1  Average orders        10  New customer  6.367796   Low spender        20.5   
2     Most orders        10  New customer  6.367796   Low spender        20.5   
3  Average orders        10  New customer  6.367796   Low spender        20.5   
4     Most orders        10  New customer  6.367796   Low spender        20.5   

          frequency_flag  
0  Non-frequent customer  
1  Non-frequent customer  
2  Non-frequent customer  
3  Non-frequent customer  
4  Non-frequent customer  

[5 rows x 25 columns]

In [10]:
# Check new memory usage

ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 25 columns):
 #   Column                 Dtype   
---  ------                 -----   
 0   order_id               int32   
 1   user_id                int32   
 2   order_number           int8    
 3   orders_day_of_week     int8    
 4   order_hour_of_day      int8    
 5   days_since_last_order  int8    
 6   product_id             int32   
 7   add_to_cart_order      int8    
 8   reordered              int8    
 9   product_name           object  
 10  aisle_id               int32   
 11  department_id          int32   
 12  prices                 float32 
 13  _merge                 category
 14  price_range_loc        object  
 15  busiest_day            object  
 16  busiest_days           object  
 17  busiest_hour           object  
 18  busiest_hours          object  
 19  max_order              int8    
 20  loyalty_flag           object  
 21  avg_price              float3

In [11]:
# Drop unneccesary columns

ords_prods_merge = ords_prods_merge.drop(columns = ['_merge','busiest_day','busiest_hour'])

In [12]:
# Check output by retrieving all column names

ords_prods_merge.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_last_order', 'product_id',
       'add_to_cart_order', 'reordered', 'product_name', 'aisle_id',
       'department_id', 'prices', 'price_range_loc', 'busiest_days',
       'busiest_hours', 'max_order', 'loyalty_flag', 'avg_price',
       'spending_flag', 'median_dslo', 'frequency_flag'],
      dtype='object')

In [ ]:
# Merge the 2 data sets

ultimate_merged_df = ords_prods_merge.merge(customer_df, on = ['user_id'], indicator = True)

### MemoryError: Unable to allocate 1.69 GiB for an array with shape (7, 32404859) and data type object
Plan B: In this plan we'll merge the 2 dataframe (with the altered data types) using a chunkwise method. This means we'll split the large data frame into chunks and merge them piece by piece to the smaller dataframe.

In [ ]:
%%time

df1 = pd.read_csv('customers.csv')
df2 = pd.read_pickle ('dataset_aggregated.pkl')
df2_key = df2['user_id']

# Creating an empty shell to save the result
df_result = pd.DataFrame(columns=(customer_df.columns.append(ords_prods_merge.columns)).unique())
df_result.to_csv(os.path.join(path,'4.3_orders_products','02 Data', 'Prepared Data', "df3.csv"),index_label=False)

# Deleting ords_prods to save memory (the large one)
del(ords_prods_merge)

# The basic idea is that we'll now load the large data set into chunks and then create a function that will 
# iteratively merge those chunks to the main data set 


# Define a function that will merge the chunks
def preprocess(x):
    df4 = pd.merge(customer_df,x, left_on = "user_id", right_on = "user_id")
    df4.to_csv(os.path.join(path,'4.3_orders_products', '02 Data', 'Prepared Data', "df3.csv"),mode="a",header=False,index=False)

    
reader = pd.read_csv(os.path.join(path,'4.3_orders_products', '02 Data', 'Original Data', 'df_large.csv'), chunksize=10000) # chunksize depends on you colsize

[preprocess(r) for r in reader]

### Plan B is not working, it keeps generating error 'no file in directory' or 'df not defined'. I don´t know what I´m doing wrong. I will export the transformed data sets as csv files and retry on another notebook.

# 05. Export Data

In [14]:
# Export the large data set

ords_prods_merge.to_pickle(os.path.join(path,'4.3_orders_products', '02 Data', 'Prepared Data', 'df_large.pkl'))

In [39]:
# Export the smaller data set

customer_df.to_csv(os.path.join(path,'4.3_orders_products', '02 Data', 'Prepared Data', 'df_small.csv'))